tracer/src/Datadog.Trace/Ci/CITracerManagerFactory.cs

```diff
@@ -34,12 +46,78 @@ internal class CITracerManagerFactory : TracerManagerFactory
 
         protected override ISampler GetSampler(ImmutableTracerSettings settings)
         {
+            if (!string.IsNullOrEmpty(_settings.ApiKey))
+            {
+                return null;
+            }
+
             return new CISampler();
         }
 
         protected override IAgentWriter GetAgentWriter(ImmutableTracerSettings settings, IDogStatsd statsd, ISampler sampler)
         {
-            return new CIAgentWriter(settings, sampler);
+            // Check for agentless scenario
+            if (_settings.Agentless)
+            {
+                if (!string.IsNullOrEmpty(_settings.ApiKey))
+                {
+                    return new CIAgentlessWriter(settings, sampler, new CIWriterHttpSender(GetRequestFactory()));
+                }
+                else
+                {
+                    Environment.FailFast("An API KEY is required in Agentless mode.");
```

Comment:
Wow this is intense. Do we need to fail so hard?

Reply:
Well, is this or an exception, due the way the testhost run tests, there's no Main method, so there's no an easy way to just exit and stop executing tests.